lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs

```diff
@@ -69,113 +66,245 @@ internal sealed class ServiceAndContextConfigurationProvider<TMapInput, TMapOutp
         /// submitted evaluator to failed evaluator
         /// </summary>
         /// <param name="evaluatorId"></param>
-        internal void EvaluatorFailed(string evaluatorId)
+        /// <returns>Whether failed evaluator is master or not</returns>
+        internal bool EvaluatorFailed(string evaluatorId)
         {
             lock (_lock)
             {
+                string msg;
+                bool isMaster = IsMaster(evaluatorId);
+
+                if (_contextLoadedEvaluators.Contains(evaluatorId))
+                {
+                    msg =
+                        string.Format(
+                            "Failed evaluator:{0} already had context loaded. Cannot handle failure at this stage",
+                            evaluatorId);
+                    Exceptions.Throw(new Exception(msg), Logger);
+                }
+
                 if (!_submittedEvaluators.Contains(evaluatorId))
                 {
-                    Exceptions.Throw(new Exception("Failed evaluator was never submitted"), Logger);
+                    msg = string.Format("Failed evaluator:{0} was never submitted", evaluatorId);
+                    Exceptions.Throw(new Exception(msg), Logger);
+                }
+
+                if (!_configurationProvider.ContainsKey(evaluatorId) && !isMaster)
+                {
+                    msg = string.Format("Partition descriptor for Failed evaluator:{0} not present", evaluatorId);
+                    Exceptions.Throw(new Exception(msg), Logger);
                 }
 
-                _failedEvaluators.Add(evaluatorId);
                 _submittedEvaluators.Remove(evaluatorId);
+
+                if (isMaster)
+                {
+                    Logger.Log(Level.Info, "Failed Evaluator is Master");
+                    _masterEvaluatorId = null;
+                    return true;
+                }
+                
+                Logger.Log(Level.Info, "Failed Evaluator is a Mapper");
+                _partitionDescriptorIds.Push(_configurationProvider[evaluatorId]);
+                _configurationProvider.Remove(evaluatorId);
+                return false;
             }
         }
 
         /// <summary>
-        /// Gives context and service configuration for next evaluator either from failed 
-        /// evaluator or new configuration
+        /// Notifies that active context state has been reached
         /// </summary>
         /// <param name="evaluatorId"></param>
-        /// <returns></returns>
-        internal ContextAndServiceConfiguration GetNextConfiguration(string evaluatorId)
+        internal void ReachedActiveContext(string evaluatorId)
```

Comment:
Agreed, let's not get into the state machine business just yet in this PR.

However, the method name needs to change, e.g. `RecordActiveContextPerEvaluatorId()`?